File: sdks/go/examples/basic_simulation.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/hypersim/hypersim-go-sdk/hypersim"
	"github.com/hypersim/hypersim-go-sdk/types"
)

func main() {
	// Initialize SDK with basic configuration
	config := &hypersim.Config{
		Network:           types.NetworkTestnet, // Use testnet for examples
		AIEnabled:         false,                // Disable AI for basic example
		CrossLayerEnabled: true,
		StreamingEnabled:  false,
		Debug:            true,
		Timeout:          30 * time.Second,
	}

	// Create SDK instance
	sdk, err := hypersim.New(config)
	if err != nil {
		log.Fatalf("Failed to create SDK: %v", err)
	}
	defer sdk.Close()

	// Create context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	// Example 1: Simple transaction simulation
	fmt.Println("=== Example 1: Simple Transaction Simulation ===")
	err = simulateTransaction(ctx, sdk)
	if err != nil {
		log.Printf("Transaction simulation failed: %v", err)
	}

	// Example 2: Get network status
	fmt.Println("\n=== Example 2: Network Status ===")
	err = getNetworkStatus(ctx, sdk)
	if err != nil {
		log.Printf("Get network status failed: %v", err)
	}

	// Example 3: Risk assessment
	fmt.Println("\n=== Example 3: Risk Assessment ===")
	err = assessTransactionRisk(ctx, sdk)
	if err != nil {
		log.Printf("Risk assessment failed: %v", err)
	}
}

func simulateTransaction(ctx context.Context, sdk *hypersim.SDK) error {
	// Create a sample transaction
	tx := &types.TransactionRequest{
		From:     "0x742d35Cc6634C0532925a3b8D9C1ddcFab6E4444",
		To:       "0xA0b86a33E6417A9cDdA68C73E2c96b4DA3AEFc43",
		Value:    "1000000000000000000", // 1 ETH in wei
		GasLimit: "21000",               // Standard gas limit for ETH transfer
		GasPrice: "20000000000",         // 20 Gwei
	}

	fmt.Printf("Simulating transaction from %s to %s\n", tx.From, tx.To)
	fmt.Printf("Value: %s wei (1 ETH)\n", tx.Value)

	// Simulate the transaction
	result, err := sdk.Simulate(ctx, tx)
	if err != nil {
		return fmt.Errorf("simulation failed: %w", err)
	}

	// Display results
	fmt.Printf("Simulation Result:\n")
	fmt.Printf("  Success: %t\n", result.Success)
	fmt.Printf("  Gas Used: %s\n", result.GasUsed)
	fmt.Printf("  Block Type: %s\n", result.BlockType)
	fmt.Printf("  Estimated Block: %d\n", result.EstimatedBlock)

	if result.Error != "" {
		fmt.Printf("  Error: %s\n", result.Error)
	}

	if result.RevertReason != "" {
		fmt.Printf("  Revert Reason: %s\n", result.RevertReason)
	}

	// Display HyperCore data if available
	if result.HyperCoreData != nil {
		fmt.Printf("  HyperCore Data:\n")
		fmt.Printf("    Positions: %d\n", len(result.HyperCoreData.Positions))
		if result.HyperCoreData.MarketData != nil {
			fmt.Printf("    Market Prices: %d\n", len(result.HyperCoreData.MarketData.Prices))
		}
	}

	return nil
}

func getNetworkStatus(ctx context.Context, sdk *hypersim.SDK) error {
	status, err := sdk.GetNetworkStatus(ctx)
	if err != nil {
		return fmt.Errorf("failed to get network status: %w", err)
	}

	fmt.Printf("Network Status:\n")
	fmt.Printf("  Network: %s\n", status.Network)
	fmt.Printf("  Latest Block: %d\n", status.LatestBlock)
	fmt.Printf("  Gas Price: %s wei\n", status.GasPrice)
	fmt.Printf("  Healthy: %t\n", status.IsHealthy)
	fmt.Printf("  Avg Block Time: %.2f seconds\n", status.AvgBlockTime)
	fmt.Printf("  Congestion Level: %s\n", status.CongestionLevel)

	return nil
}

func assessTransactionRisk(ctx context.Context, sdk *hypersim.SDK) error {
	// Create a potentially risky transaction (high gas usage)
	tx := &types.TransactionRequest{
		From:     "0x742d35Cc6634C0532925a3b8D9C1ddcFab6E4444",
		To:       "0xA0b86a33E6417A9cDdA68C73E2c96b4DA3AEFc43",
		Value:    "0",
		Data:     "0x1234567890", // Some contract call data
		GasLimit: "500000",       // Higher gas limit
		GasPrice: "50000000000",  // 50 Gwei
	}

	fmt.Printf("Assessing risk for contract interaction transaction\n")

	// Assess risk
	insights, err := sdk.AssessRisk(ctx, tx)
	if err != nil {
		return fmt.Errorf("risk assessment failed: %w", err)
	}

	fmt.Printf("Risk Assessment:\n")
	fmt.Printf("  Risk Level: %s\n", insights.RiskLevel)
	fmt.Printf("  Success Probability: %.2f\n", insights.SuccessProbability)
	fmt.Printf("  Confidence Score: %.2f\n", insights.ConfidenceScore)

	if len(insights.Recommendations) > 0 {
		fmt.Printf("  Recommendations:\n")
		for i, rec := range insights.Recommendations {
			fmt.Printf("    %d. %s\n", i+1, rec)
		}
	}

	if len(insights.SecurityWarnings) > 0 {
		fmt.Printf("  Security Warnings:\n")
		for i, warning := range insights.SecurityWarnings {
			fmt.Printf("    %d. %s\n", i+1, warning)
		}
	}

	if insights.GasOptimization != nil {
		fmt.Printf("  Gas Optimization:\n")
		fmt.Printf("    Current Gas: %s\n", insights.GasOptimization.CurrentGas)
		fmt.Printf("    Optimized Gas: %s\n", insights.GasOptimization.OptimizedGas)
		fmt.Printf("    Potential Savings: %s\n", insights.GasOptimization.Savings)
	}

	return nil
}

File: hypersim-sdk/examples/basic-simulation.ts
/**
 * Basic Transaction Simulation Example
 * 
 * This example demonstrates how to simulate a simple transaction
 * on HyperEVM and get basic analysis results.
 */

import { HyperSimSDK, Network } from '@hypersim/sdk';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

async function basicSimulationExample(): Promise<void> {
  try {
    // Initialize SDK
    const sdk = new HyperSimSDK({
      network: Network.TESTNET, // Use testnet for examples
      aiEnabled: true,
      openaiApiKey: process.env.OPENAI_API_KEY,
      debug: true
    });

    console.log('🚀 Starting basic transaction simulation...\n');

    // Example transaction: Simple ETH transfer
    const transaction = {
      from: '0x742d35Cc6635C0532925a3b8D2B9E0064c0b39ec',
      to: '0x1234567890123456789012345678901234567890',
      value: '1000000000000000000', // 1 ETH in wei
      gasLimit: '21000',
      gasPrice: '1000000000' // 1 Gwei
    };

    console.log('📝 Transaction Details:');
    console.log(`From: ${transaction.from}`);
    console.log(`To: ${transaction.to}`);
    console.log(`Value: ${transaction.value} wei (1 ETH)`);
    console.log(`Gas Limit: ${transaction.gasLimit}`);
    console.log(`Gas Price: ${transaction.gasPrice} wei (1 Gwei)\n`);

    // Simulate the transaction
    console.log('⏳ Simulating transaction...');
    const simulation = await sdk.simulate(transaction);

    console.log('\n📊 Simulation Results:');
    console.log(`Success: ${simulation.success ? '✅' : '❌'}`);
    console.log(`Gas Used: ${simulation.gasUsed}`);
    console.log(`Block Type: ${simulation.blockType}`);
    console.log(`Estimated Block: ${simulation.estimatedBlock}`);

    if (!simulation.success) {
      console.log(`❌ Error: ${simulation.error}`);
      if (simulation.revertReason) {
        console.log(`🔄 Revert Reason: ${simulation.revertReason}`);
      }
    }

    // Get AI insights if simulation was successful
    if (simulation.success && process.env.OPENAI_API_KEY) {
      console.log('\n🤖 Getting AI insights...');
      const insights = await sdk.getAIInsights(simulation);

      console.log(`\n🎯 AI Analysis:`)
      console.log(`Risk Level: ${insights.riskLevel}`);
      console.log(`Confidence: ${(insights.confidence * 100).toFixed(1)}%`);
      console.log(`\n💡 Explanation: ${insights.explanation}`);

      if (insights.gasOptimization.techniques.length > 0) {
        console.log('\n⚡ Gas Optimization Suggestions:');
        insights.gasOptimization.techniques.forEach((technique, index) => {
          console.log(`${index + 1}. ${technique.name}`);
          console.log(`   Description: ${technique.description}`);
          console.log(`   Estimated Savings: ${technique.estimatedSavings} gas`);
          console.log(`   Difficulty: ${technique.difficulty}\n`);
        });
      }

      if (insights.securityWarnings && insights.securityWarnings.length > 0) {
        console.log('🛡️ Security Warnings:');
        insights.securityWarnings.forEach((warning, index) => {
          console.log(`${index + 1}. ${warning}`);
        });
      }

      if (insights.recommendations.length > 0) {
        console.log('\n📋 Recommendations:');
        insights.recommendations.forEach((rec, index) => {
          console.log(`${index + 1}. ${rec}`);
        });
      }
    }

    // Get network status
    console.log('\n🌐 Checking network status...');
    const networkStatus = await sdk.getNetworkStatus();
    console.log(`Network: ${networkStatus.network}`);
    console.log(`Latest Block: ${networkStatus.latestBlock}`);
    console.log(`Gas Price: ${networkStatus.gasPrice} wei`);
    console.log(`Network Health: ${networkStatus.isHealthy ? '✅ Healthy' : '❌ Issues detected'}`);
    console.log(`Congestion Level: ${networkStatus.congestionLevel}`);

  } catch (error: any) {
    console.error('\n❌ Example failed:', error.message);
    console.error('Stack trace:', error.stack);
  }
}

// Run the example
if (require.main === module) {
  basicSimulationExample()
    .then(() => {
      console.log('\n✅ Basic simulation example completed successfully!');
      process.exit(0);
    })
    .catch((error) => {
      console.error('\n💥 Example failed with error:', error);
      process.exit(1);
    });
}

export { basicSimulationExample };
